ai/shared-concepts.md

```diff
@@ -131,30 +203,157 @@ let bls = BlsKeyPair::new(secret.to_string())?;
+- Use `pub mod module_name;` in `lib.rs` to expose functionality
+
+### Strict Rules
+- **MUST** use Blueprint SDK patterns for runner, router, jobs, and context
+- **MUST NOT** use `TangleConsumer`/`TangleProducer` outside Tangle-specific blueprints
+- **MUST** handle errors gracefully using `Result`
+
+---
 
+## 11. Blueprint Enforcement Rules
+
+This section consolidates all enforcement rules across Blueprint development domains.
+
+### Project Structure
+- **MUST** follow the required directory and file structure
+- **MUST** place `BlueprintRunner` setup in the binary crate's `main.rs`
+- **MUST** keep all application logic in the library crate
+- **MUST** create one module per job in the library crate's `src/jobs/`
+- **MUST** define a `Context` struct in the library crate's `src/context.rs`
+- **MUST** keep smart contract code isolated in `/contracts`
+- **MUST NOT** place logic in the binary crate besides initialization
+
+### Context and State Management
+- **MUST** include `#[config] pub env: BlueprintEnvironment` in Context
+- **MUST** derive `Clone` and necessary SDK context traits
+- **MUST** implement an `async fn new(...) -> Result<Self>` constructor
+- **MUST** initialize shared clients within the `new` function
+- **MUST** access context in handlers via the `Context<MyContext>` extractor
+
+### Job Handlers
+- **MUST** use `TangleArg`/`TangleArgsN` extractors for job inputs
+- **MUST** apply `TangleLayer` or other appropriate filters to job handlers
+- **MUST** return `TangleResult<T>` for jobs that need to report results
+- **MUST** handle errors gracefully using `Result`
+- **MUST NOT** manually decode block data; rely on extractors
+
+### Producer/Consumer
+- **MUST** use `TangleProducer` and `TangleConsumer` for Tangle interaction
+- **MUST** use `TanglePairSigner` initialized from the environment keystore
+- **MUST NOT** use `TangleConsumer`, `TangleProducer` outside Tangle-specific blueprints
+- **MUST** match Producer/Consumer to event source and target chain
+
+### Docker/Docktopus Integration
+- **MUST** use the fluent builder pattern for container creation
+- **MUST** follow the correct container lifecycle
+- **MUST** integrate container management within the Blueprint `Context`
+- **MUST** share the `Arc<Docker>` client
+- **MUST** explicitly define necessary configurations
+- **MUST NOT** rely on implicit Docker defaults
+- **MUST NOT** ignore errors from `docktopus` operations
+
+### Testing
+- **MUST** include integration tests for all job handlers
+- **MUST** use `TangleTestHarness` for Tangle Blueprints
+- **MUST** follow the setup → register → execute → verify pattern
+- **MUST** clean up resources created during tests
+- **MUST** verify both job results and any significant side effects
+
+### Error Handling
+- **MUST** propagate errors using `Result<T, E>` and `?` operator
+- **MUST NOT** use `unwrap()` or `expect()` in production code
```

Comment:
More like SHOULD NOT, unwrap and expect are perfectly valid if you verify conditions beforehand.